### PR TITLE
Show all run outputs in dedicated tab

### LINKED ISSUE #92: Show results of the entire workflow in one view
Today the user needs to click on every stop of the workflow to see it's output. This is not convenient for complex reusable workflows. We need to support a mode when the user can see the output of the entire graph in a single feed.

### LINKED ISSUE #331: 'Run output' tab for Run details page
A new tab in the Run details page called 'Run output' that lists all output produced by the pipeline. Uses the same layout as the run comparison page but shows output from a single run.

Before adding this tab, discuss with @qimingj about the concept of a pipeline output summary section 
(which contains a subset of most interesting artifacts to represent the pipelines's output/performance).

### PR DESCRIPTION
**Pull request by yebrahim**
Fixes #331.
Fixes #92.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/496)

### CODE REVIEW

File: frontend/src/lib/WorkflowParser.ts [hunk 1]
```diff
@@ -163,10 +163,17 @@ export default class WorkflowParser {
   // Returns a list of output paths for the entire workflow, by searching all nodes in
   // the workflow, and parsing outputs for each.
   public static loadAllOutputPaths(workflow: Workflow): StoragePath[] {
-    let outputPaths: StoragePath[] = [];
+    return this.loadAllOutputPathsWithStepNames(workflow).map(entry => entry[1]);
+  }
+
+  // Returns a list of object mapping a step name to output path for the entire workflow,
+  // by searching all nodes in the workflow, and parsing outputs for each.
+  public static loadAllOutputPathsWithStepNames(workflow: Workflow): Array<[string, StoragePath]> {
```
Comment (rileyjbauer) [inline frontend/src/lib/WorkflowParser.ts]: I have a slight preference for objects over tuples for this kind of thing.
For example, without coming to this function, I wouldn't know what the first string is, and in the above function `loadAllOutputPaths`, it's not immediately clear why it's safe to assume `entry[1]` is indeed a `StoragePath`.

Still, you can leave it if you prefer it

File: frontend/src/lib/WorkflowParser.ts [hunk 2]
```diff
@@ -163,10 +163,17 @@ export default class WorkflowParser {
...
+    const outputPaths: Array<[string, StoragePath]> = [];
     if (workflow && workflow.status && workflow.status.nodes) {
       Object.keys(workflow.status.nodes).forEach(n =>
-        outputPaths = outputPaths.concat(this.loadNodeOutputPaths(workflow.status.nodes[n])));
+        this.loadNodeOutputPaths(workflow.status.nodes[n]).map(path =>
```
Comment (rileyjbauer) [inline frontend/src/lib/WorkflowParser.ts]: I can't wait for a better way to iterate over values in a map like this :mask:

File: frontend/src/pages/RunDetails.test.tsx [hunk 3]
```diff
@@ -56,7 +60,10 @@ describe('RunDetails', () => {
   }
 
   beforeAll(() => jest.spyOn(console, 'error').mockImplementation());
-  afterAll(() => jest.resetAllMocks());
+  afterAll(() => {
+    jest.resetAllMocks();
```
Comment (rileyjbauer) [inline frontend/src/pages/RunDetails.test.tsx]: This is a good, but shouldn't it be `afterEach` ?

In that case, you could remove all of the `mockClear` and `mockReset`s from the `beforeEach` below.

File: frontend/src/pages/RunDetails.tsx [hunk 4]
```diff
@@ -193,14 +202,34 @@ class RunDetails extends Page<RunDetailsProps, RunDetailsState> {
...
-                {workflowParameters && !!workflowParameters.length && (<div>
-                  <DetailsTable title='Run parameters'
-                    fields={workflowParameters.map(p => [p.name, p.value || ''])} />
-                </div>)}
-              </div>}
+              {selectedTab === 1 && (
+                <div className={padding()}>
+                  {!allArtifactConfigs.length && (
+                    <span className={commonCss.absoluteCenter}>
+                      Not output artifacts found for this run.
```
Comment (rileyjbauer) [inline frontend/src/pages/RunDetails.tsx]: s/Not/No/

File: frontend/src/lib/WorkflowParser.ts (hunk 1 above)
Comment (yebrahim) [reply frontend/src/lib/WorkflowParser.ts]: No I prefer objects too. Done.

File: frontend/src/lib/WorkflowParser.ts (hunk 2 above)
Comment (yebrahim) [reply frontend/src/lib/WorkflowParser.ts]: I don't think we want an easier way. I found that lodash has `map` and `iteratee` methods that can be used here, but they don't actually look more readable.

File: frontend/src/pages/RunDetails.test.tsx (hunk 3 above)
Comment (yebrahim) [reply frontend/src/pages/RunDetails.test.tsx]: Yes it should, fixed.
Now using `jest.clearAllMocks` before each test, and `jest.resetAllMocks` after each.

Comment (rileyjbauer) [reply frontend/src/pages/RunDetails.test.tsx]: I think that might even be more than is necessary. `reset` should also clear, so you'd only be getting one extra clear called before the first test, and that likely doesn't get you anything

**Pull request merged**